lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1367 @@
+    attr->add_ints(value);
+  }
+
+  attr = proto->add_attribute();
+  attr->set_name("group");
+  attr->set_type(AttrType::INT);
+  attr->set_i(CDN->getGroup());
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeOCLConvolution(const Node *node,
+                                                 GraphType &graph,
+                                                 ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *OCLCN = llvm::dyn_cast<OCLConvolutionNode>(node);
+  RETURN_ERR_IF_NOT(OCLCN, "Invalid node type, expected OCLConvolutionNode.");
+
+  auto *attr = proto->add_attribute();
+  attr->set_name("kernel_shape");
+  attr->set_type(AttrType::INTS);
+  for (auto value : OCLCN->getKernels()) {
+    attr->add_ints(value);
+  }
+
+  attr = proto->add_attribute();
+  attr->set_name("strides");
+  attr->set_type(AttrType::INTS);
+  for (auto value : OCLCN->getStrides()) {
+    attr->add_ints(value);
+  }
+
+  attr = proto->add_attribute();
+  attr->set_name("pads");
+  attr->set_type(AttrType::INTS);
+  for (auto value : OCLCN->getPads()) {
+    attr->add_ints(value);
+  }
+
+  attr = proto->add_attribute();
+  attr->set_name("group");
+  attr->set_type(AttrType::INT);
+  attr->set_i(OCLCN->getGroup());
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeOCLAvgPool(const Node *node, GraphType &graph,
+                                             ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *OCLCN = llvm::dyn_cast<OCLAvgPoolNode>(node);
+  RETURN_ERR_IF_NOT(OCLCN, "Invalid node type, expected OCLAvgPoolNode.");
+
+  auto *attr = proto->add_attribute();
+  attr->set_name("pads");
+  attr->set_type(AttrType::INTS);
+  for (auto value : OCLCN->getPads()) {
+    attr->add_ints(value);
+  }
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+llvm::Error ONNXModelWriter::writeOCLMaxPool(const Node *node, GraphType &graph,
+                                             ReporterType &) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *OCLCN = llvm::dyn_cast<OCLMaxPoolNode>(node);
+  RETURN_ERR_IF_NOT(OCLCN, "Invalid node type, expected OCLMaxPoolNode.");
+
+  auto *attr = proto->add_attribute();
+  attr->set_name("pads");
+  attr->set_type(AttrType::INTS);
+  for (auto value : OCLCN->getPads()) {
+    attr->add_ints(value);
+  }
+
+  return writeAllWithNode(node->getName(), node, proto);
+}
+
+// Exporting arithmetic operations.
+#define DEF_ARITHMETIC_WRITER_NODE(NAME)                                       \
+  llvm::Error ONNXModelWriter::write##NAME(const Node *node, GraphType &graph, \
+                                           ReporterType &reporter) {           \
+    return writeArithmetic(#NAME, node, graph, reporter);                      \
+  }
+
+DEF_ARITHMETIC_WRITER_NODE(Add)
```

Comment:
Glow also has unary arithmetic operations. You needs to support them as well in `writeArithmetic`. Or you could have `writeBinaryArithmetic` and `writeUnaryArithmetic`.

Reply:
Can you, please, provide an example of the unary arithmetic operation Glow nodes?

Reply:
I was thinking about Log, Tanh, Sigmoid and the like.

Reply:
Log, Tanh, Sigmoid and similar operators got written by default, no special handling is required, see macro DEF_ALL_WRITER_NODE